barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp

```diff
@@ -2921,48 +2991,67 @@ AvmError AvmTraceBuilder::op_nullifier_exists(uint8_t indirect,
 
 AvmError AvmTraceBuilder::op_emit_nullifier(uint8_t indirect, uint32_t nullifier_offset)
 {
+    // We keep the first encountered error
+    AvmError error = AvmError::NO_ERROR;
     auto const clk = static_cast<uint32_t>(main_trace.size()) + 1;
 
-    auto [row, error] = create_kernel_output_opcode(indirect, clk, nullifier_offset);
+    if (nullifier_write_counter >= MAX_NULLIFIERS_PER_TX) {
+        error = AvmError::SIDE_EFFECT_LIMIT_REACHED;
+        auto row = Row{
+            .main_clk = clk,
+            .main_internal_return_ptr = internal_return_ptr,
+            .main_op_err = FF(static_cast<uint32_t>(!is_ok(error))),
+            .main_pc = pc,
+            .main_sel_op_emit_nullifier = FF(1),
+        };
+        gas_trace_builder.constrain_gas(clk, OpCode::EMITNULLIFIER);
+        main_trace.push_back(row);
+        pc += Deserialization::get_pc_increment(OpCode::EMITNULLIFIER);
+        return error;
+    }
+
+    auto [row, output_error] = create_kernel_output_opcode(indirect, clk, nullifier_offset);
     row.main_sel_op_emit_nullifier = FF(1);
+    if (is_ok(error)) {
+        error = output_error;
+    }
 
     // Do merkle check
     FF nullifier_value = row.main_ia;
     FF siloed_nullifier = AvmMerkleTreeTraceBuilder::unconstrained_silo_nullifier(
         current_public_call_request.contract_address, nullifier_value);
 
-    // This is a little bit fragile - but we use the fact that if we traced a nullifier that already exists (which is
-    // invalid), we would have stored it under a read hint.
-    NullifierReadTreeHint nullifier_read_hint = execution_hints.nullifier_read_hints.at(nullifier_read_counter);
-    bool is_update = merkle_tree_trace_builder.perform_nullifier_read(clk,
-                                                                      nullifier_read_hint.low_leaf_preimage,
-                                                                      nullifier_read_hint.low_leaf_index,
-                                                                      nullifier_read_hint.low_leaf_sibling_path,
-                                                                      intermediate_tree_snapshots.nullifier_tree.root);
+    NullifierWriteTreeHint nullifier_write_hint = execution_hints.nullifier_write_hints.at(nullifier_write_counter++);
+    bool is_update = siloed_nullifier == nullifier_write_hint.low_leaf_membership.low_leaf_preimage.next_nullifier;
```

Comment:
There is no "nullifier read" if we are doing a write. Always use the write hint, but treat it differently based on whether the siloed nullifier matches the low leaf (whether the nullifier to append is actually already present).